src/main/java/appeng/parts/misc/ItemHandlerAdapter.java

```diff
@@ -0,0 +1,362 @@
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.LinkedList;
+import java.util.Map;
+
+import net.minecraft.item.ItemStack;
+import net.minecraftforge.items.IItemHandler;
+
+import appeng.api.config.Actionable;
+import appeng.api.networking.security.BaseActionSource;
+import appeng.api.networking.storage.IBaseMonitor;
+import appeng.api.networking.ticking.TickRateModulation;
+import appeng.api.storage.IMEInventory;
+import appeng.api.storage.IMEMonitorHandlerReceiver;
+import appeng.api.storage.StorageChannel;
+import appeng.api.storage.data.IAEItemStack;
+import appeng.api.storage.data.IItemList;
+import appeng.core.AELog;
+import appeng.me.storage.ITickingMonitor;
+import appeng.util.Platform;
+import appeng.util.item.AEItemStack;
+
+
+/**
+ * Wraps an Item Handler in such a way that it can be used as an IMEInventory for items.
+ */
+class ItemHandlerAdapter implements IMEInventory<IAEItemStack>, IBaseMonitor<IAEItemStack>, ITickingMonitor
+{
+
+	private final Map<IMEMonitorHandlerReceiver<IAEItemStack>, Object> listeners = new HashMap<>();
+
+	private BaseActionSource mySource;
+
+	private final IItemHandler itemHandler;
+
+	private ItemStack[] cachedStacks = new ItemStack[0];
+
+	private IAEItemStack[] cachedAeStacks = new IAEItemStack[0];
+
+	ItemHandlerAdapter( IItemHandler itemHandler )
+	{
+		this.itemHandler = itemHandler;
+	}
+
+	@Override
+	public IAEItemStack injectItems( IAEItemStack iox, Actionable type, BaseActionSource src )
+	{
+
+		ItemStack orgInput = iox.getItemStack();
+		ItemStack remaining = orgInput;
+
+		// In simulation mode, we don't need to do 2 passes, since if we find an empty slot, we would
+		// eventually insert into it in the second phase. So for the sake of a simulation, we can consider it
+		// immediately.
+		if( type == Actionable.SIMULATE )
+		{
+			for( int i = 0; i < itemHandler.getSlots(); i++ )
+			{
+				// We have a chance to use this slot for injection
+				remaining = itemHandler.insertItem( i, remaining, true );
+				if( remaining == null )
+				{
+					break; // Awesome, full stack consumed
+				}
+			}
+		}
+		else
+		{
+
+			// For actually inserting the stack, we try to fill up existing stacks in the inventory, and then move onto free slots
+			// Think about this: In a storage drawer or barrel setup, we'd want to attempt to fill up containers that already have the
+			// item we're trying to insert first. The second phase should not cost a considerable amount of time, since it will just search
+			// for empty slots.
+			int slotCount = itemHandler.getSlots();
+
+			// This array is used to remember which slots were viable but skipped in the first phase
+			// This avoids calling getStackInSlot for each slot a second time. Hopefully the JVM will decide
+			// to allocate this array on the stack
+			boolean[] retry = new boolean[slotCount];
+
+			for( int i = 0; i < slotCount; i++ )
+			{
+				ItemStack stackInSlot = itemHandler.getStackInSlot( i );
```

Comment:
Not sure about this part. `getStackInSlot` can potentially result with an `ItemStack` being cloned, which can be quite taxing. Also the previous behaviour was to fill it simply in the order of slots. If an external inventories wants to optimise it, they can certainly try to merge stacks together on insert.